<launch>
  <!--  1) Use sim time  -->
  <param name="use_sim_time" value="true"/>

  <!--  2) How many bots? (for your own reference)  -->
  <arg name="num_bots" default="3"/>

  <!--  3) Start Gazebo with an empty world  -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name"  value="$(find gazebo_ros)/worlds/empty.world"/>
    <arg name="paused"      value="false"/>
    <arg name="use_sim_time" value="true"/>
  </include>

  <!-- Static transforms from map to odom for each robot -->
  <node pkg="tf2_ros" type="static_transform_publisher" name="map_to_odom_1" 
        args="0 0 0 0 0 0 map robot_1/odom"/>
  <node pkg="tf2_ros" type="static_transform_publisher" name="map_to_odom_2" 
        args="2 0 0 0 0 0 map robot_2/odom"/>
  <node pkg="tf2_ros" type="static_transform_publisher" name="map_to_odom_3" 
        args="4 0 0 0 0 0 map robot_3/odom"/>

  <!--  4) Spawn TurtleBot3 #1  -->
  <group ns="robot_1">
    <param 
      name="robot_description"
      command="$(find xacro)/xacro '$(find turtlebot3_description)/urdf/turtlebot3_burger.urdf.xacro'"/>
    <param name="tf_prefix" value="robot_1"/>
  </group>
  <node pkg="gazebo_ros" type="spawn_model" name="spawn_robot_1"
        args="-param /robot_1/robot_description 
              -urdf 
              -model robot_1 
              -x 0 -y 0 -z 0.01 
              -robot_namespace /robot_1"/>
  
  <rosparam ns="robot_1/diff_drive_controller">
    odom_frame_id:   'robot_1/odom'
    base_frame_id:   'robot_1/base_footprint'
    publish_rate: 50.0
    linear:
      x:
        has_velocity_limits: true
        max_velocity: 0.22
        has_acceleration_limits: true
        max_acceleration: 0.5
    angular:
      z:
        has_velocity_limits: true
        max_velocity: 2.84
        has_acceleration_limits: true
        max_acceleration: 3.2
  </rosparam>

  <!-- Publish joint states -->
  <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher_1">
    <remap from="robot_description" to="/robot_1/robot_description"/>
    <remap from="joint_states" to="/robot_1/joint_states"/>
    <param name="use_gui" value="false"/>
  </node>
  
  <!-- Publish TFs from the URDF -->
  <node pkg="robot_state_publisher"
        type="robot_state_publisher"
        name="robot_state_publisher_1"
        output="screen">
    <remap from="robot_description" to="/robot_1/robot_description"/>
    <remap from="joint_states" to="/robot_1/joint_states"/>
    <param name="tf_prefix" value="robot_1"/>
    <param name="publish_frequency" value="50.0"/>
    <param name="use_tf_static" value="true"/>
  </node>

  <!--  5) Spawn TurtleBot3 #2  -->
  <group ns="robot_2">
    <param 
      name="robot_description"
      command="$(find xacro)/xacro '$(find turtlebot3_description)/urdf/turtlebot3_burger.urdf.xacro'"/>
    <param name="tf_prefix" value="robot_2"/>
  </group>
  <node pkg="gazebo_ros" type="spawn_model" name="spawn_robot_2"
        args="-param /robot_2/robot_description 
              -urdf 
              -model robot_2 
              -x 2 -y 0 -z 0.01 
              -robot_namespace /robot_2"/>
  
  <rosparam ns="robot_2/diff_drive_controller">
    odom_frame_id:   'robot_2/odom'
    base_frame_id:   'robot_2/base_footprint'
    publish_rate: 50.0
    linear:
      x:
        has_velocity_limits: true
        max_velocity: 0.22
        has_acceleration_limits: true
        max_acceleration: 0.5
    angular:
      z:
        has_velocity_limits: true
        max_velocity: 2.84
        has_acceleration_limits: true
        max_acceleration: 3.2
  </rosparam>

  <!-- Publish joint states -->
  <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher_2">
    <remap from="robot_description" to="/robot_2/robot_description"/>
    <remap from="joint_states" to="/robot_2/joint_states"/>
    <param name="use_gui" value="false"/>
  </node>
  
  <!-- Publish TFs from the URDF -->
  <node pkg="robot_state_publisher"
        type="robot_state_publisher"
        name="robot_state_publisher_2"
        output="screen">
    <remap from="robot_description" to="/robot_2/robot_description"/>
    <remap from="joint_states" to="/robot_2/joint_states"/>
    <param name="tf_prefix" value="robot_2"/>
    <param name="publish_frequency" value="50.0"/>
    <param name="use_tf_static" value="true"/>
  </node>

  <!--  6) Spawn TurtleBot3 #3  -->
  <group ns="robot_3">
    <param 
      name="robot_description"
      command="$(find xacro)/xacro '$(find turtlebot3_description)/urdf/turtlebot3_burger.urdf.xacro'"/>
    <param name="tf_prefix" value="robot_3"/>
  </group>
  <node pkg="gazebo_ros" type="spawn_model" name="spawn_robot_3"
        args="-param /robot_3/robot_description 
              -urdf 
              -model robot_3 
              -x 4 -y 0 -z 0.01 
              -robot_namespace /robot_3"/>
 
  <rosparam ns="robot_3/diff_drive_controller">
    odom_frame_id:   'robot_3/odom'
    base_frame_id:   'robot_3/base_footprint'
    publish_rate: 50.0
    linear:
      x:
        has_velocity_limits: true
        max_velocity: 0.22
        has_acceleration_limits: true
        max_acceleration: 0.5
    angular:
      z:
        has_velocity_limits: true
        max_velocity: 2.84
        has_acceleration_limits: true
        max_acceleration: 3.2
  </rosparam>

  <!-- Publish joint states -->
  <node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher_3">
    <remap from="robot_description" to="/robot_3/robot_description"/>
    <remap from="joint_states" to="/robot_3/joint_states"/>
    <param name="use_gui" value="false"/>
  </node>
  
  <!-- Publish TFs from the URDF -->
  <node pkg="robot_state_publisher"
        type="robot_state_publisher"
        name="robot_state_publisher_3"
        output="screen">
    <remap from="robot_description" to="/robot_3/robot_description"/>
    <remap from="joint_states" to="/robot_3/joint_states"/>
    <param name="tf_prefix" value="robot_3"/>
    <param name="publish_frequency" value="50.0"/>
    <param name="use_tf_static" value="true"/>
  </node>

  <!--  7) Start your relay -->
  <include file="$(find swarm_explorer)/launch/swarm_relay.launch"/>

  <!--  8) Load swarm parameters  -->
  <rosparam file="$(find swarm_explorer)/config/swarm_params.yaml" command="load"/>

  <!--  9) Launch each explorer_bot  -->
  <group ns="robot_1">
    <include file="$(find swarm_explorer)/launch/explorer_bot.launch">
      <arg name="robot_id" value="1"/>
    </include>
  </group>
  <group ns="robot_2">
    <include file="$(find swarm_explorer)/launch/explorer_bot.launch">
      <arg name="robot_id" value="2"/>
    </include>
  </group>
  <group ns="robot_3">
    <include file="$(find swarm_explorer)/launch/explorer_bot.launch">
      <arg name="robot_id" value="3"/>
    </include>
  </group>

  <!-- Launch RViz -->
  <include file="$(find swarm_explorer)/launch/swarm_rviz.launch"/>
</launch>
